// @magicblock-labs/ephemeral-rollups-sdk/lib/__tests__/magic-router.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const magic_router_js_1 = require("../magic-router.js");
const web3_js_1 = require("@solana/web3.js");
const mockPublicKey = (address) => ({
    toBase58: () => address,
    toString: () => address,
});
jest.mock("@solana/web3.js", () => {
    const actual = jest.requireActual("@solana/web3.js");
    return {
        ...actual,
        Connection: jest.fn().mockImplementation(() => ({
            rpcEndpoint: "http://localhost",
            sendRawTransaction: jest.fn().mockResolvedValue("mock-signature"),
        })),
        Transaction: jest.fn().mockImplementation(() => ({
            feePayer: mockPublicKey("mock-fee-payer"),
            signature: [],
            instructions: [
                {
                    keys: [
                        { pubkey: mockPublicKey("key1"), isSigner: true, isWritable: true },
                        {
                            pubkey: mockPublicKey("key2"),
                            isSigner: false,
                            isWritable: false,
                        },
                    ],
                },
            ],
            serialize: jest.fn(() => Buffer.from("mock")),
            sign: jest.fn(),
        })),
        Keypair: jest.fn().mockImplementation(() => ({
            publicKey: mockPublicKey("mock-public-key"),
            sign: jest.fn(),
        })),
        PublicKey: jest
            .fn()
            .mockImplementation((address) => mockPublicKey(address)),
    };
});
global.fetch = jest.fn(async () => Promise.resolve({
    json: async () => Promise.resolve({ result: { blockhash: "mock-blockhash" } }),
}));
describe("prepareRouterTransaction", () => {
    it("sets recentBlockhash and returns the transaction", async () => {
        const connection = new web3_js_1.Connection("http://localhost");
        const transaction = new web3_js_1.Transaction();
        const result = await (0, magic_router_js_1.prepareMagicTransaction)(connection, transaction);
        expect(result.recentBlockhash).toBe("mock-blockhash");
        expect(global.fetch).toHaveBeenCalledWith("http://localhost", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method: "getBlockhashForAccounts",
                params: [["mock-fee-payer", "key1"]],
            }),
        });
    });
});
describe("sendRouterTransaction", () => {
    it("sets recentBlockhash, feePayer, signs, and sends the transaction", async () => {
        const connection = new web3_js_1.Connection("http://localhost");
        const transaction = new web3_js_1.Transaction();
        const signers = [new web3_js_1.Keypair()];
        const signature = await (0, magic_router_js_1.sendMagicTransaction)(connection, transaction, signers);
        expect(transaction.recentBlockhash).toBe("mock-blockhash");
        expect(transaction.feePayer?.toBase58()).toBe("mock-fee-payer");
        expect(transaction.sign).toHaveBeenCalledWith(...signers);
        expect(signature).toBe("mock-signature");
        expect(global.fetch).toHaveBeenCalledWith("http://localhost", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method: "getBlockhashForAccounts",
                params: [["mock-fee-payer", "key1"]],
            }),
        });
    });
});
describe("getWritableAccounts", () => {
    it("returns writable accounts from transaction", () => {
        const transaction = {
            feePayer: mockPublicKey("fee-payer"),
            instructions: [
                {
                    keys: [
                        { pubkey: mockPublicKey("key1"), isWritable: true },
                        { pubkey: mockPublicKey("key2"), isWritable: false },
                        { pubkey: mockPublicKey("key3"), isWritable: true },
                    ],
                },
            ],
        };
        const result = (0, magic_router_js_1.getWritableAccounts)(transaction);
        expect(result).toEqual(["fee-payer", "key1", "key3"]);
    });
    it("handles transaction without feePayer", () => {
        const transaction = {
            feePayer: null,
            instructions: [
                {
                    keys: [
                        { pubkey: mockPublicKey("key1"), isWritable: true },
                        { pubkey: mockPublicKey("key2"), isWritable: false },
                    ],
                },
            ],
        };
        const result = (0, magic_router_js_1.getWritableAccounts)(transaction);
        expect(result).toEqual(["key1"]);
    });
    it("handles transaction without instructions", () => {
        const transaction = {
            feePayer: mockPublicKey("fee-payer"),
            instructions: [],
        };
        const result = (0, magic_router_js_1.getWritableAccounts)(transaction);
        expect(result).toEqual(["fee-payer"]);
    });
    it("deduplicates writable accounts", () => {
        const transaction = {
            feePayer: mockPublicKey("fee-payer"),
            instructions: [
                {
                    keys: [
                        { pubkey: mockPublicKey("key1"), isWritable: true },
                        { pubkey: mockPublicKey("key1"), isWritable: true },
                        { pubkey: mockPublicKey("key2"), isWritable: false },
                    ],
                },
            ],
        };
        const result = (0, magic_router_js_1.getWritableAccounts)(transaction);
        expect(result).toEqual(["fee-payer", "key1"]);
    });
});
describe("getClosestValidator", () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });
    it("fetches and returns the closest validator public key", async () => {
        const mockIdentityData = {
            result: {
                identity: "mock-validator-identity",
            },
        };
        global.fetch.mockResolvedValueOnce({
            json: async () => Promise.resolve(mockIdentityData),
        });
        const connection = new web3_js_1.Connection("http://localhost");
        const result = await (0, magic_router_js_1.getClosestValidator)(connection);
        expect(global.fetch).toHaveBeenCalledWith("http://localhost", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method: "getIdentity",
                params: [],
            }),
        });
        expect(result.toBase58()).toBe("mock-validator-identity");
    });
    it("handles fetch errors gracefully", async () => {
        global.fetch.mockRejectedValueOnce(new Error("Network error"));
        const connection = new web3_js_1.Connection("http://localhost");
        await expect((0, magic_router_js_1.getClosestValidator)(connection)).rejects.toThrow("Network error");
    });
    it("handles invalid response format", async () => {
        global.fetch.mockResolvedValueOnce({
            json: async () => Promise.resolve({ error: "Invalid response" }),
        });
        const connection = new web3_js_1.Connection("http://localhost");
        await expect((0, magic_router_js_1.getClosestValidator)(connection)).rejects.toThrow();
    });
});
describe("getDelegationStatus", () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });
    it("returns delegation status for a string account", async () => {
        global.fetch.mockResolvedValueOnce({
            json: async () => Promise.resolve({ result: { isDelegated: false } }),
        });
        const connection = new web3_js_1.Connection("http://localhost");
        const account = "7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU";
        const result = await (0, magic_router_js_1.getDelegationStatus)(connection, account);
        expect(global.fetch).toHaveBeenCalledWith("http://localhost/getDelegationStatus", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method: "getDelegationStatus",
                params: [account],
            }),
        });
        expect(result).toEqual({ isDelegated: false });
    });
    it("returns delegation status for a PublicKey account", async () => {
        global.fetch.mockResolvedValueOnce({
            json: async () => Promise.resolve({ result: { isDelegated: true } }),
        });
        const connection = new web3_js_1.Connection("http://localhost");
        const accountKey = new web3_js_1.PublicKey("mock-public-key");
        const result = await (0, magic_router_js_1.getDelegationStatus)(connection, accountKey);
        expect(global.fetch).toHaveBeenCalledWith("http://localhost/getDelegationStatus", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method: "getDelegationStatus",
                params: [accountKey.toBase58()],
            }),
        });
        expect(result).toEqual({ isDelegated: true });
    });
});
