// @magicblock-labs/ephemeral-rollups-sdk/lib/magic-router.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWritableAccounts = getWritableAccounts;
exports.getClosestValidator = getClosestValidator;
exports.getDelegationStatus = getDelegationStatus;
exports.getLatestBlockhashForMagicTransaction = getLatestBlockhashForMagicTransaction;
exports.prepareMagicTransaction = prepareMagicTransaction;
exports.sendMagicTransaction = sendMagicTransaction;
exports.confirmMagicTransaction = confirmMagicTransaction;
exports.sendAndConfirmMagicTransaction = sendAndConfirmMagicTransaction;
exports.pollSignatureStatus = pollSignatureStatus;
const web3_js_1 = require("@solana/web3.js");
const assert_1 = __importDefault(require("assert"));
const bs58_1 = __importDefault(require("bs58"));
function getWritableAccounts(transaction) {
    const writableAccounts = new Set();
    if (transaction.feePayer) {
        writableAccounts.add(transaction.feePayer.toBase58());
    }
    for (const instruction of transaction.instructions) {
        for (const key of instruction.keys) {
            if (key.isWritable) {
                writableAccounts.add(key.pubkey.toBase58());
            }
        }
    }
    return Array.from(writableAccounts);
}
async function getClosestValidator(routerConnection) {
    const response = await fetch(routerConnection.rpcEndpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getIdentity",
            params: [],
        }),
    });
    const identityData = await response.json();
    const identity = identityData?.result?.identity;
    if (typeof identity !== "string") {
        throw new Error("Invalid identity response");
    }
    const validatorKey = {
        toBase58: () => identity,
        toString: () => identity,
    };
    return validatorKey;
}
async function getDelegationStatus(connection, account) {
    const accountAddress = typeof account === "string" ? account : account.toBase58();
    const response = await fetch(`${connection.rpcEndpoint}/getDelegationStatus`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getDelegationStatus",
            params: [accountAddress],
        }),
    });
    const data = await response.json();
    return data.result;
}
async function getLatestBlockhashForMagicTransaction(connection, transaction, options) {
    const writableAccounts = getWritableAccounts(transaction);
    const blockHashResponse = await fetch(connection.rpcEndpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getBlockhashForAccounts",
            params: [writableAccounts],
        }),
    });
    const blockHashData = await blockHashResponse.json();
    return blockHashData.result;
}
async function prepareMagicTransaction(connection, transaction, options) {
    const blockHashData = await getLatestBlockhashForMagicTransaction(connection, transaction, options);
    transaction.recentBlockhash = blockHashData.blockhash;
    return transaction;
}
async function sendMagicTransaction(connection, transaction, signersOrOptions, options) {
    const sendOpts = Array.isArray(signersOrOptions)
        ? (options ?? undefined)
        : (signersOrOptions ?? undefined);
    const latestBlockhash = await getLatestBlockhashForMagicTransaction(connection, transaction, sendOpts);
    transaction.lastValidBlockHeight =
        latestBlockhash.lastValidBlockHeight;
    transaction.recentBlockhash = latestBlockhash.blockhash;
    if (Array.isArray(signersOrOptions)) {
        transaction.sign?.(...signersOrOptions);
    }
    const wireTransaction = transaction.serialize();
    return connection.sendRawTransaction(wireTransaction, sendOpts);
}
async function confirmMagicTransaction(connection, strategy, commitment) {
    let rawSignature;
    if (typeof strategy === "string") {
        rawSignature = strategy;
    }
    else {
        const config = strategy;
        if (config.abortSignal != null && config.abortSignal.aborted) {
            return Promise.reject(config.abortSignal.reason ?? new Error("Aborted"));
        }
        rawSignature = config.signature;
    }
    let decodedSignature;
    try {
        decodedSignature = bs58_1.default.decode(rawSignature);
    }
    catch (err) {
        throw new Error("signature must be base58 encoded: " + rawSignature);
    }
    (0, assert_1.default)(decodedSignature.length === 64, "signature has invalid length");
    const status = await pollSignatureStatus(getSignatureStatus, connection, rawSignature, {
        intervalMs: 100,
        timeoutMs: 10000,
        commitment,
    });
    return status;
}
async function sendAndConfirmMagicTransaction(connection, transaction, signers, options) {
    const signature = await sendMagicTransaction(connection, transaction, signers, options);
    let status;
    if (transaction.recentBlockhash != null &&
        transaction.lastValidBlockHeight != null) {
        status = (await confirmMagicTransaction(connection, {
            abortSignal: options?.abortSignal,
            signature,
            blockhash: transaction.recentBlockhash,
            lastValidBlockHeight: transaction.lastValidBlockHeight,
        }, options?.commitment)).value;
    }
    else if (transaction.minNonceContextSlot != null &&
        transaction.nonceInfo != null) {
        const { nonceInstruction } = transaction.nonceInfo;
        const nonceAccountPubkey = nonceInstruction.keys[0].pubkey;
        status = (await confirmMagicTransaction(connection, {
            abortSignal: options?.abortSignal,
            minContextSlot: transaction.minNonceContextSlot,
            nonceAccountPubkey,
            nonceValue: transaction.nonceInfo.nonce,
            signature,
        }, options?.commitment)).value;
    }
    else {
        if (options?.abortSignal != null) {
            console.warn("sendAndConfirmTransaction(): A transaction with a deprecated confirmation strategy was " +
                "supplied along with an `abortSignal`. Only transactions having `lastValidBlockHeight` " +
                "or a combination of `nonceInfo` and `minNonceContextSlot` are abortable.");
        }
        status = (await confirmMagicTransaction(connection, signature, options?.commitment)).value;
    }
    if (status.err != null) {
        if (signature != null) {
            throw new web3_js_1.SendTransactionError({
                action: "send",
                signature,
                transactionMessage: `Status: (${JSON.stringify(status)})`,
            });
        }
        throw new Error(`Transaction ${signature} failed (${JSON.stringify(status)})`);
    }
    return signature;
}
async function getSignatureStatus(connection, signature, config) {
    const { context, value: values } = await getSignatureStatuses(connection, [signature], config);
    if (values.length === 0) {
        const value = null;
        return {
            context,
            value,
        };
    }
    else {
        (0, assert_1.default)(values.length === 1);
        const value = values[0];
        return {
            context,
            value,
        };
    }
}
async function getSignatureStatuses(connection, signatures, config) {
    const params = config ? [signatures, config] : [signatures];
    const unsafeRes = await fetch(connection.rpcEndpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "getSignatureStatuses",
            params,
        }),
    });
    const res = await unsafeRes.json();
    return res.result;
}
async function pollSignatureStatus(getSignatureStatus, connection, signature, { intervalMs = 50, timeoutMs = 12000, commitment = "confirmed", abortSignal, } = {}) {
    const maxTries = Math.ceil(timeoutMs / intervalMs);
    let tries = 0;
    return new Promise((resolve, reject) => {
        const intervalId = setInterval(() => {
            if (abortSignal != null && abortSignal.aborted) {
                clearInterval(intervalId);
                reject(abortSignal.reason ?? new Error("Polling aborted"));
                return;
            }
            tries++;
            void (async () => {
                try {
                    const result = await getSignatureStatus(connection, signature);
                    if (result.value !== null) {
                        if (result.value.confirmationStatus === commitment ||
                            result.value.confirmationStatus === "finalized") {
                            clearInterval(intervalId);
                            resolve({
                                context: result.context,
                                value: result.value,
                            });
                        }
                    }
                    else if (tries >= maxTries) {
                        clearInterval(intervalId);
                        const timeoutValue = {
                            err: new Error("Timeout"),
                        };
                        resolve({
                            context: result.context,
                            value: timeoutValue,
                        });
                    }
                }
                catch (err) {
                    clearInterval(intervalId);
                    reject(err);
                }
            })();
        }, intervalMs);
    });
}
